V2 of Auto Pokemon and TTH

madBeavis opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
Auto pokemon has more tunable parameters and can use last_seen to query data if desired.
TTH pokemon now clusters spawnpoints using Koji.

## Description
<!--- Describe your changes in detail -->
Auto:
1. How the instance operates can now be tuned by user, such as minimum spawn time, wait period after spawn, requery frequency and sleep period if algorithm gets to close to actual spawn times.
2. 60min spawns are now operational in auto.  It requires a new field in order to store the data, which is larger than just the 4bits I required for knowing if a spawn is of the 60min variety, other bits can be used in future.
3. By default 30min will be assumed, it is up to end user to likely use routes to ensure enough visits to spawnpoint to determine if it is a 30min one.
4. How 60min spawns are determined.  The hour is broken into 4 quarter hours of 15min each, ie 0-14, 15-29, 30-44 and 45-59.  If the spawn is seen in a quarter, a bit is set for it.  If the spawn has been seen in all 4 quarters, it is a 60min spawn.  If it has been seen in 3 or less, it is assumed to be 30min.  Determining such data for existing areas will require running the area in routes for a while to ensure enough samples of device being there throughout the hour are obtained.  By definition, auto alone would prove insufficient to do this as it visits spawns in generally fixed time ranges, with only a handful of spawns being determined as 60min by sheer luck of being seen by other close spawns when going to them just after spawn.

TTH:
1. Clusters using Koji, if not available falls poorboy clustering based of geohash.  For the routing, Koji returns the data sorted by points in cluster.  With data from Koji, the code starts at the biggest cluster and works its way in the direction of clusters of 1.  When requery happens, the process starts over with re-Koji of data with the new biggest cluster.
2. Requery frequency is now user adjustable
3. Added a cache to track how many devices are on instance based on user config value.  During GetTask(), devices are updated or added.  The cache will handle expiration of devices out of instance.  Low code and does not require hitting db.
4. Updated status on instance page to be far more informative, such as cluster count, unknown tth count, change of both prior, devices on instance and hops/refresh period
5. First call to get data from db and cluster using Koji is synchronous by definition.  Subsequent calls to get new data and cluster are async, so they don't block devices jumping around, they will simply go back to zero coord when new data comes in.
6. If the user sets hopTime to non-zero, the code will determine an expected amount of clusters to visit in refresh period.  With this value set, if the device(s) will visit more clusters than expected, they are returned to the start of the list.

Koji.swift (new):
1. Since this is the first use of Koji introduced into RDM, an attempt to provide a working set of code so that others may expand upon.  It works for me and I don't know others use cases, they can write their own code.
2. CodingKeys are used to keep variables in class names "Swifty" and read data properly from Koji
3. URL is defined in config by user, it is left for the code to add the specific endpoint it needs
4. I put it in "Misc" folder for the initial commit

Spawnpoint.swift:
1. Code will send a debug alert when a new TTH is found
2. Added created field to Spawnpoint.swift, so far useful for culling event spawns and will allow us to know if new spawnpoints appear in an area.
3. Added 32 bit field to track 60min spawn, which only 4 are used, can leave for future use or just make smaller.

Coord.swift:
1. Added a json decoder to Coord.swift to properly handle json from Koji.

Clamp.swift (new):
1. Added function to clamp values, simplifies code validating user config values, no repetitive and confusing min/max code cluttering things up

Geohash.swift (new):
1. Copied a small geohash library in order to use it as quick and dirty clustering as a Koji fallback

I have text for wiki and example json config example ready to be provided when requested

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
Auto:  allow the end user to config the process instead of accepting my values
TTH: Clustering using Koji speeds up the process of finding TTH
<!--- If it fixes an open issue, please link to the issue here. -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
Ran them, fixed, and now running as designed on two instances for auto.  TTH was run on one instance over period of many days and areas.
<!--- Include details of your testing environment, and the tests you ran to -->
Ran test on my bare metal server, stared at debug logs
<!--- see how your change affects other areas of the code, etc. -->
I had to add a json decoder to Coord.swift to properly handle json from Koji.  Added a debug alert to Spawnpoint.swift when new TTH is found.  Added created field to Spawnpoint.swift, so far useful for culling event spawns.  Added field to track 60min spawn, which only 4 are used, can leave for future use or just make smaller.

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!--  NOTE: In order to check code style locally and avoid having your build rejected by GitHub Actions, -->
<!--  run the following commands before you commit: `swiftlint` and `eslint`. Fix any issues they point out. -->
- [ ] My code follows the code style of this project.
- [x ] My change requires a change to the documentation.
- [x ] I have updated the documentation accordingly.

Fabio1988 commented:
Anyone except @madBeavis tested that PR? :)

madBeavis commented:
not that i am aware of, i will see if i can scrounge up at least one

Gladiator10864 commented:
I am intrigued by this PR and would like to try it but I'm not familiar enough with git to try and merge latest RDM with this PR.

Fabio1988 commented on resources/migrations/999.sql:
That's a bad name for testing... pls rename it to the proper sequence of migrations

madBeavis replied:
renamed to 100.sql

madBeavis commented on resources/migrations/999.sql:
```diff
@@ -0,0 +1,8 @@
+ALTER TABLE `spawnpoint`
+	ADD COLUMN `created` INT(11) UNSIGNED NULL DEFAULT UNIX_TIMESTAMP() AFTER `despawn_sec`,
```
renamed to stick with convention.

madBeavis replied:
edit answer: only using as a created field as it was originally named

f3ger commented:
will the merge be repeated at some point?

Fabio1988 commented:
pls resolve conflicts, update to latest.... I still did not get any feedback on this one. running at least as nice as v1?

madBeavis commented:
not aware of anyone using as i have not had any feedback of any sort, whether positive or negative.

f3ger commented:
i tested the pr, works fine and "auto tth" is faster and more efficient than in v1 due to the clustering. clustering with koji in "auto pokemon" mode would also be desirable.
I noticed a problem, which could also be me, but I didn't investigate the problem any further. when I want to display the spawn points on the map in the RDM, the RDM container crashes and restarts.

Fabio1988 commented:
> i tested the pr, works fine and "auto tth" is faster and more efficient than in v1 due to the clustering. clustering with koji in "auto pokemon" mode would also be desirable. I noticed a problem, which could also be me, but I didn't investigate the problem any further. when I want to display the spawn points on the map in the RDM, the RDM container crashes and restarts.

Thanks a lot dor your feedback.. the query on spawnpoint in map can also be related to the query adaption in this PR, would it be possible to share the error log pls? :)

Fabio1988 commented on Sources/RealDeviceMapLib/Modell/SpawnPoint.swift:
```diff
@@ -233,6 +302,8 @@ public class SpawnPoint: JSONConvertibleObject {
             let lastSeen = result[4] as! UInt32
             let despawnSecond = result[5] as? UInt16
+            let spawnInfo = result[6] as! UInt32
+            let firstSeenTimestamp = result[7] as! UInt32
```
@madBeavis unfortunately you are adding this column in DB and set it to `null` for all, but here you load them as `non-nullable` and this breaks the workflow. must this be initialized as `null` on migration or can we use `updated` timestamp instead?

madBeavis replied:
good catch, i made the changes

thanks for patience and help

Fabio1988 commented:
Not sure about latest changes. I would prefer firstSeen not to be null.. You could change the migration to use updated column as value for first seen. Would that break your algo?

madBeavis commented:
due to being 500mi away from my scanning box for another ~10days, i can't responsibly test changes to the algorithms.   crummy internet on both ends, remoting in is barely useable to keep things running.  best i can do is get home, fix any chaos that has ensued with yard and irrigation and then get onto making changes/testing.

madBeavis commented:
looking at the code, my auto only use last_seen (if user chooses the option to do such in config anyways).  for tth, the onus is on the end user to only try to find tth based off of a clean non-event set of spawnpoints.

if my memory serves me correct, i had proposed a `created` column for when the record was created with the purpose of users being able to query for range of times when event spawns were made.  up until this point, there was no functionality so far that allowed the user to do such.  i believe i was requested to change the name from `created` to `first_seen_timestamp`, i am indifferent to the name now.  i will make the change to sql file so that it doesn't default to null, which was superfluous considering the db puts in default value for now in seconds.  i can drop the idea of doing such in rdm code, but i will carry a created field in my spawnpoints table.

Fabio1988 commented:
`created` doesn't follow rdm style, fine to use `first_seen_timestamp`

madBeavis commented:
just to update fabio, i am working on issue emil007 encountered prior to getting home.  i think i found it and have a solution, so will need to do some testing on my own system before rolling it into v2.

madBeavis commented:
> Not sure about latest changes. I would prefer firstSeen not to be null.. You could change the migration to use updated column as value for first seen. Would that break your algo?

i can go with a query after alter spawnpoint table to add new fields in the migration file, ie

UPDATE spawnpoint SET first_seen_timestamp = updated;

madBeavis commented:
i got the issue emil007 found fixed.  watching two days of logs to test.  i will get ahold of emil and see if he will test the pr.

simplified much of the iterating in auto algorithm, v1 tried to be clever at the expense of readability.

kicked code to my mac, ran swiftlint and fixed issues related to my code.

madBeavis commented:
heavily revised how v2 of jumpy handles when devices get close to the actual spawn time.  put 22 devices on my test instance on area that had 3k spawnpoints.  determined that sleep in the code caused devices to drop offline as they piled up near actual spawn time, so that method of slowing things down was abandoned.  so now the code sends the device to a random spot in the list to kill time, which keeps them busy and working, so they don't fall offline.  if it is preferred, instead of random coords from area, i could send to some a few prior to current one, at least we know there are active mons there.

put 22 devices on 100+k spawnpoint area and handles the too few devices for area issue as expected.

generally reduced iteration for both situations.  my test machine didn't have issues, looking to play nice on weaker vps often used.

waiting for tester to get back to me on how things work on their end.

Fabio1988 commented:
Every time you push you are causing new linting errors 🤣 
Should I move this PR to draft state, or is it kind a complete but bug fixing?

madBeavis commented:
> Every time you push you are causing new linting errors 🤣 Should I move this PR to draft state, or is it kind a complete but bug fixing?

as long as emil can pull changes, draft is fine.

swiftlint and i don't get along.  i write code on my pc using vscode, then ftp to server to test.  once i get things right, i put onto mac and run swiftlint from there.

Emil007 commented:
I've been testing the last few iterations after encountering a bug with V1 where it (AFAIU) didn't give the workers a new job in time, which led to the devices logging out and killing everything for about 5 minutes.

After a few iterations and other strange issues, yesterday's release has been running for 16 hours now, and I don't see any mass dropouts or any other problems. The despawn  times are also **excellent** (29-27 minutes in most cases).

However, there is still something happening at the top of the hour, which leads to more activity for a short time. During the first ~10 minutes, it scans 150-300 Pokemon per device per minute, and after that, it scans 25-35 Pokemon per minute for the rest of the hour. The total Pokemon count in the database aligns with a normal Route IV scan, so it's not missing anything there. Additionally, the despawn times are very good, as I already mentioned. So, I can't complain; it's just something I wanted to note.


<img width="911" alt="image" src="https://github.com/RealDeviceMap/RealDeviceMap/assets/12276589/0f9a6385-c1fd-4d17-b5a4-4f1cbdced178">


Also, the Koji and Instance config need to be added to either the docker-compose parameters or the RDM UI. I couldn't test the Koji connection for Auto TTH because of that. It would be great to include Koji in the standard RDM settings to make it available for other future implementations. Additionally, having instance-specific settings within each instance would allow for customization and variation on a per-instance basis, and would keep Settings easier to manage instead of docker-compose

Anyway, thanks for the great work, I really like it, and i'm curious to see what linking koji will bring to the table

jfberry commented:
Remember weather changes at the top of the house, mitm behaviour could be different if it catches encounters

Fabio1988 commented:
> Also, the Koji and Instance config need to be added to either the docker-compose parameters or the RDM UI. I couldn't test the Koji connection for Auto TTH because of that. It would be great to include Koji in the standard RDM settings to make it available for other future implementations. 

Koji should be already available from local.json config? Set it there?!

Emil007 commented:
> 

you can do that when using docker-compose? Didn't know that, will look into it, then ignore the last part ;)

madBeavis commented:
koji is not used in auto pokemon, it is only used to in tth mode.  wouldn't be the first to ask about clustering in auto, but i have zero plans to add it in as auto mode is about speed.  another mitm had it, but they were bound by real world physics, where we are not so much.

madBeavis commented:
emil and i talked, i will add in a few status variables to help assign proper amount of devices to an instance.  with circles, lap times are easy to figure if they work and adjust.  with auto, it has no such capability other than watching debug logs.  so code will track resets of position in coordinate array from too many devices or too few devices.  there will be a small amount of false positives, but if one instance has 100 too fast and the others have 5 over an hour, one can say there are too many devices on that area.

madBeavis commented:
![image](https://github.com/RealDeviceMap/RealDeviceMap/assets/57059024/d23295f5-16f8-4c20-9534-ba23f37c2d90)

here is sample status for auto mode, will give idea to end user if they have too many or too few devices in order to help efficiently use devices.  cleaned up status for tth mode.

Fabio1988 commented:
@madBeavis good for review? You have updated wiki description for those instances?

madBeavis commented:
Auto Pokemon
scans IV for nearby (~70m) Pokemon for coordinates in a given geofence specified in scan area based off of spawn times

- Name: A unique name of the Instance.
- Timezone offset: Not used.
- Min Level: Minimum worker level to be used. Set to 30 to get the proper IV.
- Max Level: Maximum worker level to be used.
- Scan Area: Geofence to query for spawnpoints
- Account Group: Group from which an account will be pulled for this instance.
- Is Unique Event: implemented for ticketed events. When activated, spawns will be separated from standard spawns.
- Note: This spawn mode is for those that require better spawn times remaining on pokemon than the other circle pokemon modes. The circle modes can cover more spawnpoints per hour at the expense of timer quality. This mode gives the best timers on spawns at the cost of assigning more devices to the instance. You may assign multiple devices to this instance, they will leapfrog over each other.
- Notes for configuration variables:  You will need to setup docker to use the local.json config.  The variables of this mode are under /scanModes/auto
1. useLastSeenTime: Setting this value determines how the mode queries the db in respect to last_seen time.  This value is in seconds.  To use all spawnpoints in the area with a tth without regard to last seen, set it to -1.  To query spawnpoints only seen in the last hour, set the value to 3600.
2. requeryFrequency: Determines how often the mode queries the db for spawnpoints, defaults to 3600 seconds.
3. minSpawnTime: The value of minimum spawn times the mode will consider valid in seconds.  If you want minimum 25 minute timers, set to 1800.  The mode will not send devices to spawns of 1757 seconds.  If this situation occurs, the mode will shift the working window of valid spawns.
4. bufferTime: How close to actual spawn time that the mode will visit the spot in seconds.  This value allows for vagaries in the whole system, such as network delays.  The default value is 30 seconds, so for a 30 minute spawn, the mode will not visit the spot before 29:30.
5. sleepInterval:  How close in seconds that the mode can get to the spawn time minus buffer time.  Say you set a buffer time of 30 seconds.  If you have too many devices on an instance, it will approach the above 29:30 and at 29:25 it will send devices to random recent spots to keep them from over running the desired window.
6. defaultLongitude: String value reserved for future use
7. defaultLatitude: String value reserved for future use

- Default is Note on spawn time lengths: Given enough visits to a spawnpoint during portions of the hour, it can be determined if the spawn is a 30 or 60 minute timer.  Some may occur while finding tth.  By the nature of how the mode works, the device will only visit right after the monster has spawned and will not get a good distribution of visits throughout the hour to determine if it is a 60 minute spawn.  Some information may be obtained if it is viewed by nearby spots.  To be 100% sure if the spawnpoint is of the 30 or 60 minute variety, you may run circle pokemon over the area to spread visits to the point around the hour.
- Notes for setup and tuning: For first run, assign one device so that you can get a count of spawnpoints in the geofence or use Koji. For calculating the amount of devices refer to your seconds/jump. With that information you can calculate the maximal amount of spawnpoints covered in an hour assuming they are evenly distributed. For 3s/jump it's recommended to use 900-1000 spawnpoints/device to leave you a buffer.  If you run too many devices on an instance, they will be sent to random recent coords to keep busy and not timeout.  If you run too few devices on an instance, they will fall behind and the devices will skip over coords to return to spawns within the time range you desire based on values
- Notes on random timers:  Even though a 60 minute timer may not be known by the db, based on when the spawn is found and when it disappears, the map and alerting software will correctly identify it with greater than 30 minutes left.  You may also see some low timers, this is to be expected if you run too few devices, causing the mode to skip spawns that then get picked up on a visit to a nearby spot.
- Notes on status:  V2 has been upgraded to give more informative status information.  If you have too many devices on an instance, it will tell how many times devices have been sent to random coords.  If you have too few devices on an instance, it will tell how many times the valid time window has been set and how many spawnpoints have been skipped.




Auto TTH
Determine TTH (time till hidden) for spawnpoints in a given geofence

- Name: A unique name of the Instance.
- Timezone offset: Not used.
- Min Level: Minimum worker level to be used. Set to 30 to get the proper IV.
- Max Level: Maximum worker level to be used.
- Scan Area: List of geofence to query for spawnpoints that need tth found. You can use multiple geofences for this instance.
- Account Group: Group from which an account will be pulled for this instance.
- Is Unique Event: implemented for ticketed events. When activated, spawns will be separated from standard spawns.
- Note: This mode is intended narrow down find tth for spawnpoints in a given geofence. As designed, you give it geofence and it works through them until it runs out of coords without tth.  You may assign multiple devices to this instance, they will leapfrog over each other.
- Notes for initialization: Be sure to properly bootstrap (initialize) an area. At a mininum, you want three evenly timed passes at 0min, 20min and 40min past the top of the hour. Since that proposition is difficult, I like to do more passes than 3, but you do you. If your spawns are close to each other, you can get away with a crummy init as they will get found during the process of finding tth for other points, then get worked in during the next grind during the next requery.
- Notes for configuration variables:  You will need to setup docker to use the local.json config.  The variables of this mode are under /scanModes/tth
1. requeryFrequency: This is how often the db is requeried for tth changes and new spots.  Default is 90s, which is the period of tth.
2. clusterUsingKoji: Does the mode use Koji for clustering.  If this variable is false, the mode will use basic geohash clustering as a slight improvement over using the raw coordinates for points.  If set to true, the code will use Koji, which is a vast improvement over all other methods.  The raw coordinates obtained will be sent to Koji, clustered and returned.  Devices will start with the densest clusters first and go down in density.
3. clusteringRadius: What raidius is used for clustering.
4. deviceTimeout: Timer to keep track of how many devices are active on the instance, if the device has not been seen in more than set seconds, it is considered offline
5. hopTime: Used in status calculations
6. Further tuning: Monitor the status messages on the instance page.  The mode will output information to help you monitor the situation.

madBeavis commented:
all known issues with my code are resolved.  ready for review and merge.
